crates/ra_mbe/src/syntax_bridge.rs

```diff
@@ -119,47 +149,102 @@ impl TtTokenSource {
         res.convert_subtree(tt);
         res
     }
+
     fn convert_subtree(&mut self, sub: &tt::Subtree) {
         self.push_delim(sub.delimiter, false);
-        sub.token_trees.iter().for_each(|tt| self.convert_tt(tt));
+
+        let mut pos = 0;
+        while pos < sub.token_trees.len() {
+            pos += self.convert_tt(pos, sub);
```

Comment:
this `pos` handling seems complicated. I think this is because in `convert_tt` we sort-of try to convert an token, but actually may convert several at once. I think that, instead of back and forth between these two functions, the code will become clearer if we lift some of the logic from convert_tt to this function.

Something like this:

```
let it = sub.token_tree.siter().peek();
while let Some(t) = it.next() {
    match t {
       // use peek here to advance the inter by more than one token
    }
}
```

Reply:
Good idea, i will implement it in the join `::` PR later.